File: app/src/main/java/com/divercity/app/features/signup/SignUpViewModel.java
package com.divercity.app.features.signup;

import com.divercity.app.core.base.BaseViewModel;
import com.divercity.app.core.utils.SingleLiveEvent;
import com.divercity.app.data.Resource;
import com.divercity.app.data.entity.signup.response.SignUpResponse;
import com.divercity.app.features.signup.usecase.SignUpUseCase;

import javax.inject.Inject;

/**
 * Created by lucas on 27/09/2018.
 */

public class SignUpViewModel extends BaseViewModel {

    private SignUpUseCase signUpUseCase;

    private SingleLiveEvent<Resource<Void>> signUp = new SingleLiveEvent<>();

    @Inject
    public SignUpViewModel(SignUpUseCase signUpUseCase) {
        this.signUpUseCase = signUpUseCase;
    }

    public void signUp(String name, String username, String email, String password, String confirmPassword) {
        signUp.setValue(Resource.loading(null));
        SignUpUseCase.Callback callback = new SignUpUseCase.Callback() {
            @Override
            protected void onFail(String msg) {
                signUp.setValue(Resource.error(msg, null));
            }

            @Override
            protected void onSuccess(SignUpResponse response) {
                signUp.setValue(Resource.success(null));
            }
        };
        getCompositeDisposable().add(callback);
        signUpUseCase.execute(callback, SignUpUseCase.Params.forSignUp(
                username,
                name,
                email,
                password,
                confirmPassword));
    }

    public SingleLiveEvent<Resource<Void>> getSignUp() {
        return signUp;
    }

}

File: app/src/main/java/com/divercity/app/features/signup/SignUpViewModel.kt
package com.divercity.app.features.signup

import com.divercity.app.core.base.BaseViewModel
import com.divercity.app.core.utils.SingleLiveEvent
import com.divercity.app.data.Resource
import com.divercity.app.data.entity.login.response.LoginResponse
import com.divercity.app.data.networking.config.DisposableObserverWrapper
import com.divercity.app.features.signup.usecase.CheckIsUsernameRegisteredUseCase
import com.divercity.app.features.signup.usecase.SignUpUseCase
import com.divercity.app.features.signup.usecase.UploadProfilePictureUseCase
import com.google.gson.JsonElement
import javax.inject.Inject

/**
 * Created by lucas on 27/09/2018.
 */

class SignUpViewModel @Inject
constructor(private val signUpUseCase: SignUpUseCase,
            private val usernameRegisteredUseCase: CheckIsUsernameRegisteredUseCase,
            private val uploadProfilePictureUseCase: UploadProfilePictureUseCase) : BaseViewModel() {

    val signUpResponse = SingleLiveEvent<Resource<Any>>()
    val usernameRegisteredResponse = SingleLiveEvent<Resource<Boolean>>()
    val uploadProfilePictureResponse = SingleLiveEvent<Resource<LoginResponse>>()
    val navigateToSelectUserType = SingleLiveEvent<Boolean>()

    fun signUp(name: String, username: String, email: String, password: String, confirmPassword: String) {

        signUpResponse.value = Resource.loading(null)
        val callback = object : SignUpUseCase.Callback() {
            override fun onFail(msg: String) {
                signUpResponse.value = Resource.error(msg, null)
            }

            override fun onSuccess(response: LoginResponse) {
                signUpResponse.value = Resource.success(response)
            }
        }
        compositeDisposable.add(callback)
        signUpUseCase.execute(callback, SignUpUseCase.Params.forSignUp(
                username,
                name,
                email,
                password,
                confirmPassword))
    }

    fun checkUsernameRegistered(username: String) {

        val callback = object : DisposableObserverWrapper<Boolean>() {

            override fun onFail(error: String) {
            }

            override fun onHttpException(error: JsonElement?) {
            }

            override fun onSuccess(t: Boolean) {
                usernameRegisteredResponse.value = Resource.success(t)
            }
        }
        compositeDisposable.add(callback)
        usernameRegisteredUseCase.execute(callback, CheckIsUsernameRegisteredUseCase.Params.forCheckUsername(username))
    }

    fun uploadPicture(pictureBase64: String) {
        if (pictureBase64 == "")
            navigateToSelectUserType.call()
        else {
            val callback = object : DisposableObserverWrapper<LoginResponse>() {

                override fun onFail(error: String) {
                    uploadProfilePictureResponse.value = Resource.error(error, null)
                }

                override fun onHttpException(error: JsonElement?) {

                }

                override fun onSuccess(t: LoginResponse) {
                    uploadProfilePictureResponse.value = Resource.success(t)
                }
            }
            compositeDisposable.add(callback)
            uploadProfilePictureUseCase.execute(callback, UploadProfilePictureUseCase.Params.forUploadPic(pictureBase64))
        }
    }
}
